cogito/put.go

```diff
@@ -136,21 +138,67 @@ func (putter *ProdPutter) LoadConfiguration(in io.Reader, args []string) error {
 }
 
 func (putter *ProdPutter) ProcessInputDir() error {
-	inputDirs, err := collectInputDirs(putter.InputDir)
+	// putter.InputDir, corresponding to key "put:inputs:", should contain 1 or 2 dirs.
+	// If it contains one, we support autodiscovery by not requiring to name it, we know
+	// that it should be the git repo.
+	// If on the other hand it contains two, one should be the git repo (still nameless)
+	// and the other should be the directory containing the chat_message_file, which is
+	// named by the first element of the path in "chat_message_file".
+	// This allows (although clumsily) to distinguish which is which.
+	// This complexity has historical reasons to preserve backwards compatibility
+	// (the nameless git repo).
+	//
+	// Somehow independent is the reason why we enforce the count of directories to be
+	// max 2: this is to avoid the default Concourse behavior of streaming _all_ the
+	// volumes "just in case".
```

Comment:
This explains why the code is convoluted. I am happy if we find a simpler approach.

Reply:
Personally, I would have done something like this (pseudo code)
```
inputDirs = collectInputDirs(putter.InputDir)
if params.ChatMessageFile != "" {
  msgDir, _ = path.Split(params.ChatMessageFile)
  msgDir = strings.TrimSuffix(msgDir, "/")
  if msgDir == "" {
    return fmt.Errorf("blabla wrong format")
  }
  _, found = inputDirs.pop(msgDir)
  if !found {
    return fmt.Errorf("blabla not found")
  }
}
if inputDirs.size() != 1 {
  return fmt.Errorf("blabla missing git repo")
}
```
But I'm a bit biased by the feature rich sets in Python. Maybe you can still pop from a map in Go?

Even without "pop" you could still avoid checking `ChatMessageFile` twice. If it's there, you first check if you have 2 inputs then check one of them hold the chat message file. Otherwise you just check you've got a single input.

Reply:
Thinking a bit more about it, if you Go has lists and trying to remove stuff from it can signal if the element wasn't found, that's enough to implement the same algo and then you wouldn't need a Set type.

Reply:
For sure you can pop from a Go map, even in assembly you can pop :-D

Thanks for the suggestion, let me go through it.

Reply:
@odormond with a fresh mind I now think I understand your proposal. Seems nice indeed. Will try it, thanks.

Reply:
@odormond the last part
```
if inputDirs.size() != 1 {
  return fmt.Errorf("blabla missing git repo")
}
```
was not enough. I had to change it to:
```
	if inputDirs.Size() == 0 {
		return fmt.Errorf(
			"put:inputs: missing directory for GitHub repo: have: %v, GitHub: %s/%s",
			collected, source.Owner, source.Repo)
	} else if inputDirs.Size() > 1 {
		return fmt.Errorf(
			"put:inputs: want only directory for GitHub repo: have: %v, GitHub: %s/%s",
			inputDirs, source.Owner, source.Repo)
	}
```

It might still be more readable than what I had.